// @soroswap/sdk/dist/types/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedProtocols = exports.SupportedNetworks = exports.SupportedPlatforms = exports.SupportedAssetLists = exports.TradeType = void 0;
// Trade types
var TradeType;
(function (TradeType) {
    TradeType["EXACT_IN"] = "EXACT_IN";
    TradeType["EXACT_OUT"] = "EXACT_OUT";
})(TradeType || (exports.TradeType = TradeType = {}));
// Asset list types
var SupportedAssetLists;
(function (SupportedAssetLists) {
    SupportedAssetLists["SOROSWAP"] = "https://raw.githubusercontent.com/soroswap/token-list/main/tokenList.json";
    SupportedAssetLists["STELLAR_EXPERT"] = "https://api.stellar.expert/explorer/public/asset-list/top50";
    SupportedAssetLists["LOBSTR"] = "https://lobstr.co/api/v1/sep/assets/curated.json";
    SupportedAssetLists["AQUA"] = "https://amm-api.aqua.network/tokens/?format=json&pooled=true&size=200";
})(SupportedAssetLists || (exports.SupportedAssetLists = SupportedAssetLists = {}));
var SupportedPlatforms;
(function (SupportedPlatforms) {
    SupportedPlatforms["SDEX"] = "sdex";
    SupportedPlatforms["AGGREGATOR"] = "aggregator";
    SupportedPlatforms["ROUTER"] = "router";
})(SupportedPlatforms || (exports.SupportedPlatforms = SupportedPlatforms = {}));
var SupportedNetworks;
(function (SupportedNetworks) {
    SupportedNetworks["TESTNET"] = "testnet";
    SupportedNetworks["MAINNET"] = "mainnet";
})(SupportedNetworks || (exports.SupportedNetworks = SupportedNetworks = {}));
var SupportedProtocols;
(function (SupportedProtocols) {
    SupportedProtocols["SOROSWAP"] = "soroswap";
    SupportedProtocols["PHOENIX"] = "phoenix";
    SupportedProtocols["AQUA"] = "aqua";
    // COMET = 'comet',
    SupportedProtocols["SDEX"] = "sdex";
})(SupportedProtocols || (exports.SupportedProtocols = SupportedProtocols = {}));
